### DISCUSSION TITLE
GMSH import mesh file

### DISCUSSION THREAD

**Discussion by lolsenkettle** [Chaste/Chaste#254] [Q&A] [answered]
Can a GMSH (https://gmsh.info/) mesh file be imported into CHASTE Cardiac? Will the Physical Surfaces remain tagged if it is exported into VTK file format?

Thanks!

**Comment by jmpf** [comment]
There is a GMSH reader.  It doesn't look as if the Cardiac is set up to use it though.  It looks from https://chaste.github.io/doxygen-latest/GenericMeshReader_8hpp_source.html like the code will try to read in in Tetgen/Triangle format, Memfem format and VTK format, in that order.

I suggest converting the mesh before using Chaste.  However we might be able to experiment with your data and see whether a small change to Chaste can enable it to import or convert.

**Reply by lolsenkettle** [reply]
Thanks, I will try using tetgen first. I am just starting to use CHASTE but have used gmsh before so wanted to ask if it could be imported.

**Comment by lolsenkettle** [comment]
I have changed the code for the tutorial SolvingLinearPdesTutorial to read in a gmsh file in the directory below called Cube.msh:

        GmshMeshReader<3,3> mesh_reader("projects/BionicEye/test/GmshCube/Cube");

        /* Now declare a tetrahedral mesh with the same dimensions... */
        TetrahedralMesh<3,3> mesh;
        /* ... and construct the mesh using the mesh reader. */
        /* mesh.ConstructFromGmshMeshReader(mesh_reader); */
        mesh.ConstructFromMeshReader(mesh_reader);

I also include the line #include "GmshMeshReader.hpp" at top of file.

However when I run the code it gives the following error below. Could you let me know if this is how to read in a Gmsh file?

Consolidate compiler generated dependencies of target chaste_project_BionicEye
[100%] Linking CXX shared library libchaste_project_BionicEye.so
[100%] Built target chaste_project_BionicEye
[100%] Generating TestSolvingLinearPdesTutorial3.cpp

No tests defined



make[3]: *** [projects/BionicEye/test/CMakeFiles/TestSolvingLinearPdesTutorial3.dir/build.make:74: projects/BionicEye/test/TestSolvingLinearPdesTutorial3.cpp] Error 2
make[2]: *** [CMakeFiles/Makefile2:22430: projects/BionicEye/test/CMakeFiles/TestSolvingLinearPdesTutorial3.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:22438: projects/BionicEye/test/CMakeFiles/TestSolvingLinearPdesTutorial3.dir/rule] Error 2
make: *** [Makefile:8428: TestSolvingLinearPdesTutorial3] Error 2

**Comment by jmpf** [comment]
I guess you have correctly done these steps:
- Copied test `cp pde/test/tutorials/TestSolvingLinearPdesTutorial.hpp projects/MyProject/test/TestSolvingLinearPdesTutorial3.hpp`
- Added to the new test suite project's ContinuousTestPack.txt
- Made the change outlined above

However "No tests defined" in the text above is an alert that you have also changed the names of the test functions in that test suite.  Only functions which match the names  `void Test...` are recognised and run by the testing framework.  You may sometimes see deactivated tests in suites that look like `void xTest...` but you are not allowed to deactivate them all.

**Reply by lolsenkettle** [reply]
Yes I have done those steps. I also copied GmshMeshReader.hpp into my project directory. I added both TestSolvingLinearPdesTutorial3.hpp and GmshMeshReader.hpp to ContinuousTestPack.txt.  

I also tried running it without GmshMeshReader.hpp in my project directory and removed that line from ContinuousTestPack.txt. But I still get the same error:

[100%] Built target chaste_project_BionicEye
[100%] Generating TestSolvingLinearPdesTutorial3.cpp

No tests defined

make[3]: *** [projects/BionicEye/test/CMakeFiles/TestSolvingLinearPdesTutorial3.dir/build.make:74: projects/BionicEye/test/TestSolvingLinearPdesTutorial3.cpp] Error 2
make[2]: *** [CMakeFiles/Makefile2:22430: projects/BionicEye/test/CMakeFiles/TestSolvingLinearPdesTutorial3.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:22438: projects/BionicEye/test/CMakeFiles/TestSolvingLinearPdesTutorial3.dir/rule] Error 2
make: *** [Makefile:8428: TestSolvingLinearPdesTutorial3] Error 2

**Reply by jmpf** [reply]
> Yes I have done those steps. I also copied GmshMeshReader.hpp into my project directory. I added both TestSolvingLinearPdesTutorial3.hpp and GmshMeshReader.hpp to ContinuousTestPack.txt.

You don't need to copy any files from the trunk into your project.

**Reply by jmpf** [reply]
> > Yes I have done those steps. I also copied GmshMeshReader.hpp into my project directory. I added both TestSolvingLinearPdesTutorial3.hpp and GmshMeshReader.hpp to ContinuousTestPack.txt.
> 
> You don't need to copy any files from the trunk into your project.

I mean nothing for trunk **source**.  TestSolvingLinearPdesTutorial has relies on 2 files which are in the test folder with it, so you also need to copy `pde/test/pdes/HeatEquationWithSourceTerm.hpp` and `pde/test/pdes/SimplePoissonEquation.hpp`

Did you rerun cmake after deleting GmshReader.hpp from the ContinuousTestPack.txt?

I cannot reproduce your error locally.  Perhaps you could show us the code?  Please either make these changes on a public repo or attach to this issue.

**Reply by lolsenkettle** [reply]
I already had pde/test/pdes/HeatEquationWithSourceTerm.hpp and pde/test/pdes/SimplePoissonEquation.hpp in my test folder to get the tutorial to work without making any other changes. I'm attaching my code and gmsh files. I tried with and without the .msh and for both Cube.msh and CubePhysical.msh files created in gmsh. 
[send.zip](https://github.com/user-attachments/files/16971236/send.zip)

**Reply by jmpf** [reply]
> I already had pde/test/pdes/HeatEquationWithSourceTerm.hpp and pde/test/pdes/SimplePoissonEquation.hpp in my test folder to get the tutorial to work without making any other changes. I'm attaching my code and gmsh files. I tried with and without the .msh and for both Cube.msh and CubePhysical.msh files created in gmsh. [send.zip](https://github.com/user-attachments/files/16971236/send.zip)

Thanks for the attachment.  Now I can see what is going on.   ContinuousTestPack.txt is all about telling the cmake build system where to find the test files which need to be built into executable.  Having HeatEquationWithSourceTerm.hpp and SimplePoissonEquation.hpp named there means that it tries (and fails) to build them separately.

- Delete the last two lines from ContinuousTestPack.txt
- Rerun the cmake command: `cmake ..` (or similar)
- Recompile: `make project_myprojectname` (or similar)
- Run the executable: 'ctest  -VV -R TestSolvingLinearPdesTutorial3`

**Reply by lolsenkettle** [reply]
Unfortunately removing those lines didn't make any difference and I get the same error:

 96%] Built target chaste_project_BionicEye
[100%] Generating TestSolvingLinearPdesTutorial3.cpp

No tests defined

make[3]: *** [projects/BionicEye/test/CMakeFiles/TestSolvingLinearPdesTutorial3.dir/build.make:74: projects/BionicEye/test/TestSolvingLinearPdesTutorial3.cpp] Error 2
make[2]: *** [CMakeFiles/Makefile2:22422: projects/BionicEye/test/CMakeFiles/TestSolvingLinearPdesTutorial3.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:22430: projects/BionicEye/test/CMakeFiles/TestSolvingLinearPdesTutorial3.dir/rule] Error 2
make: *** [Makefile:8428: TestSolvingLinearPdesTutorial3] Error 2

**Comment by jmosborne** [comment]
Did you run cmake config again?

cmake ../src

or similar

**Reply by lolsenkettle** [reply]
Yes I do that every time and run configure twice and generate once. Also it is worth noting that the tutorial using the tutorial without changes: TestSolvingLinearPdesTutorial runs with or without the 2 files (HeatEquationWithSourceTerm.hpp and SimplePoissonEquation.hpp): in the ContinuousTestPack.txt

**Comment by jmosborne** [comment]
The only thing that should be in ContinuousTestPack.txt should be the one
test you have in the test folder TestSolvingLinearPdesTutorial3.hpp

**Reply by lolsenkettle** [reply]
Yes still same error with only this file in ContinuousTestPack.txt

**Reply by lolsenkettle** [reply, marked as answer]
I'm sorry but I think the problem is my end when I tried to remove the elliptic PDE problem. I just tried copying the tutorial file again and leaving everything alone except read mesh and now it works perfectly! Sorry for this!

**Comment by jmosborne** [comment]
All good, glad it works!
